File: src/main/java/uk/co/threebugs/mochiwhattotrade3/setupgroup/SetupGroupService.java
package uk.co.threebugs.mochiwhattotrade3.setupgroup;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.AllArgsConstructor;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.stereotype.Service;
import uk.co.threebugs.mochiwhattotrade3.setupgroups.SetupGroupsFile;
import uk.co.threebugs.mochiwhattotrade3.setupgroups.SetupGroupsMapper;
import uk.co.threebugs.mochiwhattotrade3.setupgroups.SetupGroupsRepository;

import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.stream.Collectors;

@AllArgsConstructor
@Service
public class SetupGroupService {

    private static final Logger logger = LogManager.getLogger(SetupGroupService.class);

    private final SetupGroupMapper setupGroupMapper;
    private final SetupGroupRepository setupGroupRepository;
    private final SetupGroupsRepository setupGroupsRepository;
    private final SetupGroupsMapper setupGroupsMapper;
    private final ObjectMapper objectMapper;

    public List<SetupGroup> loadSetupsFromFile(Path setupsPath) {
        try {
            var setupGroupsFile = objectMapper.readValue(setupsPath.toFile(), SetupGroupsFile.class);
            var scriptsDirectory = Paths.get(setupGroupsFile.getScriptsDirectory());

            var setupGroups = setupGroupsRepository.findByName(setupGroupsFile.getName())
                                                   .orElseGet(() -> setupGroupsRepository.save(setupGroupsMapper.toEntity(setupGroupsFile)));

            return setupGroupsFile.getSetupGroups()
                                  .stream()
                                  .map(setupGroup -> SetupGroupDto.builder()
                                                                  .path(scriptsDirectory.resolve(setupGroup.getPath()))
                                                                  .symbol(setupGroup.getSymbol())
                                                                  .enabled(setupGroup.getEnabled())
                                                                  .build())
                                  .map(setupGroupDto -> {

                                      var optionalSetupGroup = setupGroupRepository.findByPath(setupGroupDto.getPath()
                                                                                                            .toString());

                                      return optionalSetupGroup.orElseGet(() -> {
                                          var setupGroup = setupGroupMapper.toEntity(setupGroupDto, setupGroups);
                                          return setupGroupRepository.save(setupGroup);
                                      });
                                  })
                                  .collect(Collectors.toList());
        } catch (IOException e) {
            logger.error("Error loading setups: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/uk/co/threebugs/darwinexclient/setupgroup/SetupGroupService.kt
package uk.co.threebugs.darwinexclient.setupgroup

import com.fasterxml.jackson.databind.*
import org.springframework.stereotype.*
import uk.co.threebugs.darwinexclient.setupgroups.*
import uk.co.threebugs.darwinexclient.utils.*
import java.io.*
import java.nio.file.*
import java.util.stream.*

@Service
class SetupGroupService(
    private val setupGroupMapper: SetupGroupMapper,
    private val setupGroupRepository: SetupGroupRepository,
    private val setupGroupsRepository: SetupGroupsRepository,
    private val setupGroupsMapper: SetupGroupsMapper,
    private val objectMapper: ObjectMapper
) {
    fun loadSetupsFromFile(setupsPath: Path): List<SetupGroup> {
        return try {
            val setupGroupsFile = objectMapper.readValue(setupsPath.toFile(), SetupGroupsFile::class.java)
            val scriptsDirectory = Paths.get(setupGroupsFile.scriptsDirectory)
            val setupGroups = setupGroupsRepository.findByName(setupGroupsFile.name)
                .orElseGet { setupGroupsRepository.save(setupGroupsMapper.toEntity(setupGroupsFile)) }
            setupGroupsFile.setupGroups
                .stream()
                .map { setupGroup: SetupGroupFile ->
                    SetupGroupDto(
                        path = scriptsDirectory.resolve(setupGroup.path),
                        symbol = setupGroup.symbol,
                        enabled = setupGroup.enabled,
                        direction = setupGroup.direction,
                        setupGroups = setupGroupsMapper.toDto(setupGroups)
                    )
                }
                .map { setupGroupDto: SetupGroupDto ->
                    val optionalSetupGroup = setupGroupRepository.findByPath(
                        setupGroupDto.path
                            .toString()
                    )
                    optionalSetupGroup.orElseGet {
                        val setupGroup = setupGroupMapper.toEntity(setupGroupDto, setupGroups)
                        setupGroupRepository.save(setupGroup)
                    }
                }
                .collect(Collectors.toList())
        } catch (e: IOException) {
            logger.error("Error loading setups: " + e.message)
            throw RuntimeException(e)
        }
    }
}
